File: mahaam-api-cs/Src/Infra/Security.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security;
using System.Security.Claims;
using System.Text;
using Mahaam.Feat.Users;
using Microsoft.IdentityModel.Tokens;

namespace Mahaam.Infra;

public interface IAuth
{
	(Guid, Guid, bool) ValidateAndExtractJwt(HttpContext context);
	string CreateToken(string userId, string deviceId);
}

public class Auth(IDeviceRepo deviceRepo, IUserRepo userRepo, Settings settings) : IAuth
{
	public (Guid, Guid, bool) ValidateAndExtractJwt(HttpContext context)
	{
		string? authorization = context.Request.Headers.Authorization;
		if (string.IsNullOrEmpty(authorization))
		{
			throw new UnauthorizedException($"Authorization header not exists");
		}
		string tokenString = authorization[7..]; // Remove 'Bearer ' to get the jwt

		ValidateJwt(tokenString);
		var token = new JwtSecurityToken(tokenString);


		var userId = GetGuidClaim(token, "userId");
		var deviceId = GetGuidClaim(token, "deviceId");

		var device = deviceRepo.GetOne(deviceId).GetAwaiter().GetResult();
		if ((device is null || userId != device.UserId) && !context.Request.Path.Equals("/user/logout"))
			throw new UnauthorizedException($"Invalid user info");

		var user = userRepo.GetOne(userId).GetAwaiter().GetResult();
		var isLoggedIn = user is not null && user.Email is not null;
		return (userId, deviceId, isLoggedIn);
	}

	private static Guid GetGuidClaim(JwtSecurityToken token, string claimType)
	{
		var claim = token.Claims.First(claim => claim.Type == claimType);
		Rule.Required(claim.Value, claimType);

		if (!Guid.TryParse(claim.Value, out Guid id) || id == Guid.Empty)
		{
			throw new ForbiddenException($"{claimType} is empty");
		}
		return id;
	}

	public string CreateToken(string userId, string deviceId)
	{
		try
		{
			var creds = new SigningCredentials(SecurityKey(), SecurityAlgorithms.HmacSha256);
			var descriptor = new SecurityTokenDescriptor()
			{
				Subject = new ClaimsIdentity([
					new Claim("userId", userId),
					new Claim("deviceId", deviceId),
				]),
				Expires = DateTime.Now.Add(TimeSpan.FromDays(7)),
				SigningCredentials = creds,
				IssuedAt = DateTime.Now,
				Issuer = "mahaam-api",
			};
			var handler = new JwtSecurityTokenHandler();
			var token = handler.CreateToken(descriptor);
			return handler.WriteToken(token);
		}
		catch (Exception exception)
		{
			throw new SecurityException(exception.ToString());
		}
	}

	public void ValidateJwt(string token)
	{
		try
		{
			var tokenHandler = new JwtSecurityTokenHandler();
			var validationParams = GetValidationParams();
			tokenHandler.ValidateToken(token, validationParams, out _);
		}
		catch (SecurityTokenExpiredException)
		{
			throw new UnauthorizedException("Token has expired");
		}
		catch (SecurityTokenException)
		{
			throw new UnauthorizedException("Invalid token");
		}
		catch (Exception)
		{
			throw new UnauthorizedException("Token validation failed");
		}
	}

	private TokenValidationParameters GetValidationParams()
	{
		return new TokenValidationParameters()
		{
			ValidateLifetime = true,
			ValidateAudience = false,
			ValidateIssuer = true,
			ValidIssuer = "mahaam-api",
			IssuerSigningKey = SecurityKey()
		};
	}


	private SymmetricSecurityKey SecurityKey()
	{
		var keyBytes = Encoding.ASCII.GetBytes(settings.Api.TokenSecretKey);
		return new SymmetricSecurityKey(keyBytes);
	}
}

File: mahaam-api-java/src/main/java/mahaam/infra/Security.java
package mahaam.infra;

import java.nio.charset.StandardCharsets;
import java.security.Key;
import java.util.Date;
import java.util.UUID;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import io.jsonwebtoken.security.Keys;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.container.ContainerRequestContext;
import mahaam.feat.user.DeviceRepo;
import mahaam.feat.user.UserModel.User;
import mahaam.feat.user.UserRepo;
import mahaam.infra.Exceptions.ForbiddenException;
import mahaam.infra.Exceptions.UnauthorizedException;

@ApplicationScoped
public class Security {

	@Inject
	DeviceRepo deviceRepo;

	@Inject
	UserRepo userRepo;

	@Inject
	Config config;

	public static class AuthResult {
		private UUID userId;
		private UUID deviceId;
		private boolean isLoggedIn;

		public AuthResult(UUID userId, UUID deviceId, boolean isLoggedIn) {
			this.userId = userId;
			this.deviceId = deviceId;
			this.isLoggedIn = isLoggedIn;
		}

		public UUID getUserId() {
			return userId;
		}

		public UUID getDeviceId() {
			return deviceId;
		}

		public boolean isLoggedIn() {
			return isLoggedIn;
		}
	}

	private static final String EMPTY_UUID = "00000000-0000-0000-0000-000000000000";

	public static void nonEmptyUuid(String uuidString, String name) {
		Rule.required(uuidString, name);
		if (uuidString == null || uuidString.trim().isEmpty() || EMPTY_UUID.equals(uuidString)) {
			throw new ForbiddenException(name + " is Empty");
		}
	}

	public AuthResult validateAndExtractJwt(ContainerRequestContext context)
			throws UnauthorizedException, ForbiddenException {
		String path = context.getUriInfo().getPath();
		String authorization = context.getHeaderString("Authorization");

		if (authorization == null || authorization.isEmpty()) {
			throw new UnauthorizedException("Authorization header not exists");
		}

		if (!authorization.startsWith("Bearer ")) {
			throw new UnauthorizedException("Invalid Authorization header format");
		}

		String tokenString = authorization.substring(7); // Remove 'Bearer ' to get the jwt

		try {
			validate(tokenString);
			Claims claims = Jwts.parserBuilder()
					.setSigningKey(getSecurityKey())
					.build()
					.parseClaimsJws(tokenString)
					.getBody();

			String userIdClaim = claims.get("userId", String.class);
			nonEmptyUuid(userIdClaim, "userId");

			String deviceIdClaim = claims.get("deviceId", String.class);
			nonEmptyUuid(deviceIdClaim, "deviceId");

			UUID userId = UUID.fromString(userIdClaim);
			UUID deviceId = UUID.fromString(deviceIdClaim);

			// Use the device repository to get device info
			var device = deviceRepo.getOne(deviceId);
			if ((device == null || !userId.equals(device.userId)) && !"/user/logout".equals(path)) {
				throw new UnauthorizedException("Invalid user info");
			}

			User user = userRepo.getOne(userId);
			boolean isLoggedIn = user.email != null;

			return new AuthResult(userId, deviceId, isLoggedIn);

		} catch (Exception e) {
			if (e instanceof UnauthorizedException || e instanceof ForbiddenException) {
				throw e;
			}
			throw new UnauthorizedException("Invalid JWT token: " + e.getMessage());
		}
	}

	public String createToken(String userId, String deviceId) {
		try {
			Key key = getSecurityKey();
			Date now = new Date();
			Date expiration = new Date(now.getTime() + (7 * 24 * 60 * 60 * 1000)); // 7 days

			return Jwts.builder()
					.claim("userId", userId)
					.claim("deviceId", deviceId)
					.setIssuedAt(now)
					.setExpiration(expiration)
					.setIssuer("mahaam-api")
					.signWith(key, SignatureAlgorithm.HS256)
					.compact();

		} catch (Exception e) {
			throw new RuntimeException("Error creating JWT token: " + e.getMessage(), e);
		}
	}

	private void validate(String token) {
		try {
			Jwts.parserBuilder()
					.setSigningKey(getSecurityKey())
					.requireIssuer("mahaam-api")
					.build()
					.parseClaimsJws(token);
		} catch (Exception e) {
			throw new RuntimeException("JWT validation failed: " + e.getMessage(), e);
		}
	}

	private Key getSecurityKey() {
		byte[] keyBytes = config.tokenSecretKey().getBytes(StandardCharsets.UTF_8);
		return Keys.hmacShaKeyFor(keyBytes);
	}

}
